Add a view display extender for the localgov_page_header for the lede

andybroomfield opened the pull request:
Fix #127
<!-- See https://docs.localgovdrupal.org/contributing/ for guidelines on contributing. -->

## What does this change?

Adds a setting to views that allows for a page header category, with the option
to set a custom lede inside the view, which will then be used by the
page header block.

<!-- A pull request should have enough detail to be understandable far in the
future. e.g what is the problem/why is the change needed, how does it solve it
and any questions or points of discussion. Prefer copying information from a
Trello card (for example) over linking to it; the card may not always exist and
reviewers may not have access to the board. -->

## How to test
1. Enable the new page header display extender on /admin/structure/views/settings/advanced (should we enable this by default?)
2. Create a new view with a page display.
3. There should be a new page header section in the center of the views configuration.
4. Edit that and add some summary text, you should also be able to use tokens from the view.
5. View the views page, and the page header should also display the entered summary as the lede (part of the page header block).

<!-- Provide instructions to help others verify the change. This could take the
form of "On main, do X and witness Y. On this branch, do X and witness Z. " -->

## How can we measure success?

<!-- Do you expect errors to decrease? Do you expect user journeys to be
simplified? What can be used to prove this? A filtered view of logs or
analytics, etc? -->
Now possible for site builders to add subtitles to views.

## Have we considered potential risks?

<!-- What are the potential risks and how can they be mitigated? Does an error
require an alarm? Should user help, infosec, or legal be informed of this
change? Is private information guarded? Do we need to add anything in the
backlog? -->

May conflict with others who have added their own extension to get a subtitle on the view.

## Images

<!-- Usually only applicable to UI changes, what did it look like before and
what will it look like after? -->

<img width="1232" alt="Screenshot 2024-11-13 at 9 47 10 am" src="https://github.com/user-attachments/assets/c0fd2e34-3bdb-4502-87d1-49a58c633c1a">

<img width="474" alt="Screenshot 2024-11-13 at 9 52 19 am" src="https://github.com/user-attachments/assets/bbe8d984-c582-4c66-a6df-fb5850bfc1ce">

<img width="1048" alt="Screenshot 2024-11-13 at 9 52 23 am" src="https://github.com/user-attachments/assets/347a6aa5-9582-4cab-895f-d11d0ad9f791">


## Accessibility

<!-- Usually only applicable to UI changes, check the boxes if you are satisfied that your changes pass these tests -->

-   [ ] [Tested with screen reader](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#screen-reader)
-   [ ] [Navigable with keyboard](https://github.com/guardian/accessibility/blob/main/people-and-technology/02-physical.md#keyboard)
-   [ ] [Colour contrast passed](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#contrast)
-   [ ] [The change doesn't use only colour to convey meaning](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#use-of-colour)

andybroomfield committed:
Add a view display extender for the localgov_page_header for the lede

Fix #127

Adds a setting to views that allows for a page header category, with the option
to set a custom lede inside the view, which will then be used by the
page header block.

andybroomfield committed:
Correct page header display extender title

andybroomfield commented on src/Plugin/Block/PageHeaderBlock.php:
```diff
@@ -152,6 +155,13 @@ public function __construct(array $configuration, $plugin_id, $plugin_definition
+        $view = Views::getView($defaults['view_id']);
+        if ($view) {
+          $view->setDisplay($defaults['display_id']);
+          $this->entity = $view;
```
Not of type Drupal\Core\Entity\EntityInterface. 
The pageHeaderEvent isn't doing type checking (maybe it should) so perhaps we need to send $view separately?

finnlewis commented:
@Adnan-cds can see this will be potentially useful!

andybroomfield committed:
Add seperate view property with setter and getter to pageHeaderDisplayEvent

Since a viewExecutable is not a child of EntityInterface, provide seperate
$view property and set that when viewing a view page from the pageHeaderBlock.

andybroomfield committed:
Add comment on why $view has to be rendered and cs fixes

Adnan-cds commented:
It works!  Thanks Andy :)

The code looks okay at first glance.  I would suggest that Steve @stephen-cox does a thorough review.

finnlewis commented:
Some questions from Merge Tuesday: 

1. Do we want to enable this by default?
2. How do we enable this by default? (Follow how MetaTag does this?)
3. Code review: happy with the approach?

stephen-cox commented on src/Plugin/views/display_extender/PageHeaderDisplayExtender.php:
```diff
@@ -0,0 +1,251 @@
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    /** @var \Drupal\metatag_views\Plugin\views\display_extender\MetatagDisplayExtender */
```
This type hint references the Metatag display extender and isn't needed here anyway

stephen-cox reviewed (commented):
I'm happy with this approach. One small nitpick with a type hint that's wrong and can probably be removed.

This does add a hard dependency between Core and Views. I don't have a problem with this as I can't think of an LGD site that wouldn't use Views, but we should add the dependency in the info.yml file.

I think it would also be good to enable this by default and maybe add an update hook to enable it for others. The test already includes the code on how to do this.

https://github.com/localgovdrupal/localgov_core/blob/c9c7ad9cafa1096e425c08e31c1fc38d3cb8560a/tests/src/Functional/ViewsPageExtenderTest.php#L40-L44

andybroomfield committed:
Fix create docblock event in views display extender

andybroomfield committed:
Add views as dependency as pageDisplayExtender requires it

andybroomfield committed:
Enable the views page header display extender by default

- Enables by default and adds update hook.
- Removes the install from the test.

andybroomfield committed:
Fix missing return type errors

stephen-cox commented:
@stephen-cox Will test the update hook and merge if okay

andybroomfield commented:
Discussed on MT on 28/1/2025: We'd like to get this reviewed soon. @ekes @stephen-cox please.

stephen-cox reviewed (approved):
Looking good to me :+1:
